cookbooks/bcpc-hadoop/libraries/utils.rb

```diff
@@ -203,59 +203,84 @@ def set_hosts
     'hbase_head' => 'role[BCPC-Hadoop-Head-HBase]'
   }
 
+  service_recipe = {
+    'oozie_server' => 'recipe[bcpc-hadoop::oozie]'
+  }
+
   if node.run_state['cluster_def'].nil? then
     node.run_state['cluster_def'] = BACH::ClusterDef.new(node_obj: node)
   end
-  cd = node.run_state['cluster_def']
+  nodes = node.run_state['cluster_def'].fetch_cluster_def
+
+  # host search helper lambdas
+  runs_role = Proc.new { |host, role| host[:runlist].include? service_role[role] }
+  runs_recipe = Proc.new { |host, recipe| host[:runlist].include? service_recipe[recipe] }
+
+  # mapped host objects
+  to_host = Proc.new do |host| {
+    'hostname' => host[:fqdn],
+    'node_number' => host[:node_id],
+    'zookeeper_myid' => nil
+  } end
   
   node.default[:bcpc][:hadoop][:zookeeper][:servers] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['zookeeper'] }.map {
-      |hst| { 'hostname' => hst[:fqdn], 'node_number' => hst[:node_id], 'zookeeper_myid' => nil } }
+    nodes.select { |n| runs_role.call(n, 'zookeeper') }.map { |n| to_host.call(n) }
```

Comment:
Can you iterate over a list of roles here rather than hardcoding them? (Ideally can you introspect the roles to see which ones include Oozie?)

Reply:
Yea, I can probably arrange these individual statements into a clever loop. I'll make that change.